// @dcloudio/uni-mp-compiler/dist/template/codegen.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.genElementProps = exports.genNode = exports.generate = void 0;
const shared_1 = require("@vue/shared");
const uni_shared_1 = require("@dcloudio/uni-shared");
const uni_cli_shared_1 = require("@dcloudio/uni-cli-shared");
const compiler_core_1 = require("@vue/compiler-core");
const codegen_1 = require("../codegen");
const vFor_1 = require("../transforms/vFor");
const vIf_1 = require("../transforms/vIf");
const vSlot_1 = require("../transforms/vSlot");
const utils_1 = require("../transforms/utils");
/**
 * 注意此处的 escapeText 并未解决用户代码内的实体字符与产物内的不一致的Bug。
 * vue编译器在tokenize阶段会将实体字符转义为对应的字符，因此在codegen阶段无法做到完美还原用户代码。
 * 但是在uni-app-x依然要做反转义，主要考虑以下几点：
 * - 用户源码&gt;原产物为>，导致wxml解析错误
 * - 用户源码内不会出现>字符。虽然emsp等字符可以出现在用户的源码里面，但是一般不会有人这么做。因此无论用户写的是&emsp;还是\u2003，都被转义为&emsp;对用户而言影响不大
 */
const mpEscapeText = (0, uni_cli_shared_1.getEscaper)(/[<>\u2009\u00A0\u2002\u2003]/g, new Map([
    [60, '&lt;'],
    [62, '&gt;'],
    [0x2009, '&thinsp;'],
    [0xa0, '&nbsp;'],
    [0x2002, '&ensp;'],
    [0x2003, '&emsp;'],
]));
function generate({ children }, { slot, event, scopeId, emitFile, filename, directive, lazyElement, isBuiltInComponent, isMiniProgramComponent, checkPropName, component, autoImportFilters, filter, isX, }) {
    const context = {
        slot,
        event,
        code: '',
        scopeId,
        directive,
        lazyElement,
        component,
        isBuiltInComponent,
        isMiniProgramComponent,
        checkPropName,
        push(code) {
            context.code += code;
        },
        isX,
    };
    children.forEach((node) => {
        genNode(node, context);
    });
    if (filter && filter.generate && autoImportFilters.length) {
        autoImportFilters.forEach((autoImportFilter) => {
            context.code +=
                filter.generate(autoImportFilter, (process.env.UNI_SUBPACKAGE
                    ? `/${process.env.UNI_SUBPACKAGE}/common/`
                    : '/common/') + autoImportFilter.id) + '\n';
        });
    }
    emitFile({ type: 'asset', fileName: filename, source: context.code });
}
exports.generate = generate;
function isInVFor(node) {
    while (node) {
        if ((0, vFor_1.isForElementNode)(node)) {
            return true;
        }
        node = node.parent;
    }
    return false;
}
function genNode(node, context) {
    switch (node.type) {
        case compiler_core_1.NodeTypes.IF:
            return node.branches.forEach((node) => {
                genNode(node, context);
            });
        case compiler_core_1.NodeTypes.TEXT:
            return genText(node, context);
        case compiler_core_1.NodeTypes.INTERPOLATION:
            return genExpression(node.content, context);
        case compiler_core_1.NodeTypes.ELEMENT:
            if (node.tagType === compiler_core_1.ElementTypes.SLOT) {
                const isEmptyDefaultSlot = node.props.some((p) => ((0, uni_cli_shared_1.isAttributeNode)(p) &&
                    p.name === 'name' &&
                    p.value?.content === uni_shared_1.SLOT_DEFAULT_NAME) ||
                    (p.name === 'bind' &&
                        p.slotName ===
                            uni_shared_1.SLOT_DEFAULT_NAME)) && node.children.length === 0;
                // 当存在 <slot name="default" :xxx="xxx"><slot> 时，在后面添加 <slot></slot>，使默认插槽生效
                if (isEmptyDefaultSlot) {
                    if (isInVFor(node)) {
                        return genSlot(node, context);
                    }
                    const isVIfSlot = (0, vIf_1.isIfElementNode)(node);
                    if (isVIfSlot) {
                        context.push(`<block`);
                        genVIfCode(node, context);
                        context.push(`>`);
                        delete node.vIf;
                    }
                    genSlot(node, context);
                    genSlot((0, shared_1.extend)({}, node, { props: [], children: [], loc: {} }), context);
                    if (isVIfSlot) {
                        context.push(`</block>`);
                    }
                    return;
                }
                return genSlot(node, context);
            }
            else if (node.tagType === compiler_core_1.ElementTypes.COMPONENT) {
                return genComponent(node, context);
            }
            else if (node.tagType === compiler_core_1.ElementTypes.TEMPLATE) {
                return genTemplate(node, context);
            }
            else if (isLazyElement(node, context)) {
                return genLazyElement(node, context);
            }
            return genElement(node, context);
    }
}
exports.genNode = genNode;
function genText(node, { push, isX }) {
    if (isX) {
        push(mpEscapeText(node.content));
    }
    else {
        // 目前暂时只处理 < 和 >，防止微信小程序编译报错 ask 138888
        push((0, uni_cli_shared_1.getEscaper)(/[<>]/g, new Map([
            [60, '&lt;'],
            [62, '&gt;'],
        ]))(node.content));
    }
}
function genExpression(node, { push }) {
    push(`{{${(0, codegen_1.genExpr)(node)}}}`);
}
function genVIf(exp, { push, directive }) {
    push(` ${directive}if="{{${exp}}}"`);
}
function genVElseIf(exp, { push, directive }) {
    push(` ${directive}elif="{{${exp}}}"`);
}
function genVElse({ push, directive }) {
    push(` ${directive}else`);
}
function genVFor(node, { push, directive }) {
    const { sourceCode, valueAlias, indexAlias } = node.vFor;
    push(` ${directive}for="${sourceCode}"`);
    if (valueAlias) {
        push(` ${directive}for-item="${valueAlias}"`);
    }
    if (valueAlias === 'index') {
        push(` ${directive}for-index="${indexAlias}"`);
    }
    const keyProp = (0, compiler_core_1.findProp)(node, 'key', true);
    if (keyProp) {
        const key = keyProp.exp.content;
        push(` ${directive}key="${key.includes('.') ? key.split('.')[1] : key}"`);
        node.props.splice(node.props.indexOf(keyProp), 1);
    }
}
function genSlot(node, context) {
    // 移除掉所有非name属性，即移除作用域插槽的绑定指令
    node.props = node.props.filter((prop) => {
        if ((0, uni_cli_shared_1.isAttributeNode)(prop)) {
            return prop.name === 'name';
        }
        else if (prop.arg?.type === compiler_core_1.NodeTypes.SIMPLE_EXPRESSION) {
            return prop.arg.content === 'name';
        }
    });
    const isDefaultSlot = node.props.some((p) => (0, uni_cli_shared_1.isAttributeNode)(p) &&
        p.name === 'name' &&
        p.value?.content === uni_shared_1.SLOT_DEFAULT_NAME);
    if (!node.children.length ||
        (context.slot.fallbackContent && !isDefaultSlot)) {
        // 无后备内容或支持后备内容
        return genElement(node, context);
    }
    const { push } = context;
    const isVIfSlot = (0, vIf_1.isIfElementNode)(node);
    if (isVIfSlot) {
        push(`<block`);
        genVIfCode(node, context);
        push(`>`);
        delete node.vIf;
    }
    const children = node.children.slice();
    node.children.length = 0;
    push(`<block`);
    const nameProp = (0, compiler_core_1.findProp)(node, 'name');
    let name = uni_shared_1.SLOT_DEFAULT_NAME;
    if (nameProp) {
        if ((0, uni_cli_shared_1.isAttributeNode)(nameProp)) {
            if (nameProp.value?.content) {
                name = nameProp.value.content;
            }
        }
        else {
            if (nameProp.slotName) {
                name = nameProp.slotName;
            }
        }
    }
    if (name.includes('-') || /^\d/.test(name)) {
        genVIf(`$slots['${name}']`, context);
    }
    else {
        genVIf(`$slots.${name}`, context);
    }
    push(`>`);
    genElement(node, context);
    // 当存在 <slot name="default" :xxx="xxx"> fallback <slot> 时，在后面添加 <slot></slot>，使默认插槽生效
    if (isDefaultSlot) {
        push(`<slot/>`);
    }
    push(`</block>`);
    push(`<block`);
    genVElse(context);
    push(`>`);
    // 默认插槽 且支持 fallback，fallback 需要包裹在 <slot> 中
    if (context.slot.fallbackContent && isDefaultSlot) {
        push(`<slot>`);
    }
    children.forEach((node) => {
        genNode(node, context);
    });
    if (context.slot.fallbackContent && isDefaultSlot) {
        push(`</slot>`);
    }
    push(`</block>`);
    if (isVIfSlot) {
        push(`</block>`);
    }
}
function genTemplate(node, context) {
    const slotProp = node.props.find((prop) => (0, uni_cli_shared_1.isDirectiveNode)(prop) &&
        (prop.name === 'slot' ||
            (prop.name === 'bind' &&
                prop.arg?.type === compiler_core_1.NodeTypes.SIMPLE_EXPRESSION &&
                prop.arg.content === 'slot')));
    // 为 bind 时，通常是作用域插槽生成的 vSlot.ts:197 createBindDirectiveNode('slot',...)
    if (slotProp && (slotProp.name === 'bind' || (0, vSlot_1.findSlotName)(slotProp))) {
        /**
         * 仅百度、字节支持使用 block 作为命名插槽根节点
         * 此处为了统一仅默认替换为view
         * <template v-slot/> => <view slot="">
         */
        node.tag = 'view';
    }
    else {
        // <template/> => <block/>
        node.tag = 'block';
    }
    // @ts-expect-error
    node.tagType = compiler_core_1.ElementTypes.ELEMENT;
    // 仅单个子节点的命名插槽(非作用域)，直接使用子节点作为插槽使用，避免多增加的 view 节点影响 flex 排版
    if (slotProp &&
        node.tag === 'view' &&
        !(0, vFor_1.isForElementNode)(node) &&
        node.children.length === 1) {
        const child = node.children[0];
        if ((0, uni_cli_shared_1.isElementNode)(child) &&
            !(0, vFor_1.isForElementNode)(child) &&
            !(0, compiler_core_1.isSlotOutlet)(child)) {
            child.props.push(slotProp);
            if ((0, vIf_1.isIfElementNode)(node)) {
                ;
                child.vIf = node.vIf;
            }
            return genElement(child, context);
        }
    }
    else if (slotProp && node.tag === 'view' && context.isX) {
        /**
         * uni-app-x小程序端为了对齐app平台view设置了默认的overflow: hidden样式
         * 对于slot生成的view节点，这个默认样式大多情况下不符合开发者预期
         * 在此view节点补充style="overflow: visible"
         */
        node.props.push((0, uni_cli_shared_1.createAttributeNode)('style', 'overflow: visible'));
        /**
         * 补充flex相关样式，解决部分场景下slot内容无法正确flex布局的问题
         */
        node.props.push((0, uni_cli_shared_1.createAttributeNode)('class', 'uni__inherit_flex_box_style'));
    }
    return genElement(node, context);
}
function genComponent(node, context) {
    if (context.component?.getPropertySync) {
        return genElement(node, context);
    }
    if ((0, vIf_1.isIfElementNode)(node) || (0, vFor_1.isForElementNode)(node)) {
        return genElement(node, context);
    }
    // 小程序原生组件，补充 if(r0)
    if (context.isMiniProgramComponent(node.tag)) {
        ;
        node.vIf = {
            name: 'if',
            condition: 'r0',
        };
        return genElement(node, context);
    }
    const prop = (0, compiler_core_1.findProp)(node, utils_1.ATTR_VUE_PROPS);
    if (!prop) {
        return genElement(node, context);
    }
    ;
    node.vIf = {
        name: 'if',
        condition: prop.exp.content,
    };
    return genElement(node, context);
}
function isLazyElement(node, context) {
    if (!context.lazyElement) {
        return false;
    }
    let lazyProps;
    if ((0, shared_1.isFunction)(context.lazyElement)) {
        const res = context.lazyElement(node, context);
        if (!(0, shared_1.isPlainObject)(res)) {
            return res;
        }
        lazyProps = res[node.tag];
    }
    else {
        lazyProps = context.lazyElement[node.tag];
    }
    if (lazyProps === true) {
        return true;
    }
    if (!lazyProps) {
        return;
    }
    return node.props.some((prop) => (0, uni_cli_shared_1.isDirectiveNode)(prop) &&
        lazyProps.find((lazyProp) => {
            return (prop.name === lazyProp.name &&
                prop.arg?.type === compiler_core_1.NodeTypes.SIMPLE_EXPRESSION &&
                lazyProp.arg.includes(prop.arg.content));
        }));
}
/**
 * 部分内置组件的部分事件在初始化时会立刻触发，但标准事件需要等首次渲染才能确认事件函数，故增加wx:if="{{r0}}"
 * @param node
 * @param context
 */
function genLazyElement(node, context) {
    const { push } = context;
    if (!(0, vIf_1.isIfElementNode)(node)) {
        push(`<block`);
        // r0 => ready 首次渲染
        genVIf(`r0`, context);
        push(`>`);
        genElement(node, context);
        push(`</block>`);
        return;
    }
    // v-if,v-else-if 无需处理
    if (node.vIf.name !== 'else') {
        return genElement(node, context);
    }
    push(`<block`);
    genVElse(context);
    push(`>`);
    node.vIf.name = 'if';
    node.vIf.condition = 'r0';
    genElement(node, context);
    push(`</block>`);
}
function genVIfCode(node, context) {
    const { name, condition } = node.vIf;
    if (name === 'if') {
        genVIf(condition, context);
    }
    else if (name === 'else-if') {
        genVElseIf(condition, context);
    }
    else if (name === 'else') {
        genVElse(context);
    }
}
function genElement(node, context) {
    const { children, isSelfClosing, props } = node;
    let tag = node.tag;
    // <template slot="left"/> => <block slot="left"/>
    if (tag === 'template') {
        if ((0, compiler_core_1.findProp)(node, 'slot')) {
            tag = 'view';
        }
        else {
            tag = 'block';
        }
    }
    // 无用的 block
    if (tag === 'block' &&
        props.length === 0 &&
        !(0, vIf_1.isIfElementNode)(node) &&
        !(0, vFor_1.isForElementNode)(node)) {
        return children.forEach((node) => {
            genNode(node, context);
        });
    }
    let virtualHost = false;
    if ((0, uni_cli_shared_1.isUserComponent)(node, context)) {
        tag = (0, shared_1.hyphenate)(tag);
        if (context.component?.normalizeName) {
            tag = context.component?.normalizeName(tag);
        }
        if (context.component?.mergeVirtualHostAttributes) {
            virtualHost = true;
        }
    }
    const { push } = context;
    const hasVIf = (0, vIf_1.isIfElementNode)(node);
    const hasVFor = (0, vFor_1.isForElementNode)(node);
    const hasVIfAndVFor = hasVIf && hasVFor;
    // 小程序中 wx:else wx:elif 不支持与 wx:for 同时使用
    // 故 if 需要补充一层 block
    if (hasVIfAndVFor) {
        push(`<block`);
        genVIfCode(node, context);
        push(`>`);
    }
    push(`<${tag}`);
    if (!hasVIfAndVFor && hasVIf) {
        genVIfCode(node, context);
    }
    if (hasVFor) {
        genVFor(node, context);
    }
    if (props.length) {
        genElementProps(node, virtualHost, context);
    }
    if (isSelfClosing) {
        push(`/>`);
    }
    else {
        push(`>`);
        children.forEach((node) => {
            genNode(node, context);
        });
        push(`</${tag}>`);
    }
    if (hasVIfAndVFor) {
        push(`</block>`);
    }
}
function checkVirtualHostProps(name, virtualHost) {
    const names = [name];
    if (virtualHost) {
        const obj = {
            style: uni_shared_1.VIRTUAL_HOST_STYLE,
            class: uni_shared_1.VIRTUAL_HOST_CLASS,
            hidden: uni_shared_1.VIRTUAL_HOST_HIDDEN,
            id: uni_shared_1.VIRTUAL_HOST_ID,
        };
        if (name in obj) {
            // TODO 支付宝平台移除原有属性（支付宝小程序自定义组件外部属性始终无效）
            names.push(obj[name]);
        }
        return names;
    }
    return names;
}
function genElementProps(node, virtualHost, context) {
    node.props.forEach((prop) => {
        if ((0, uni_cli_shared_1.isAttributeNode)(prop)) {
            if (context.checkPropName &&
                !context.checkPropName(prop.name, prop, node)) {
                return;
            }
            const { value } = prop;
            if (value) {
                checkVirtualHostProps(prop.name, virtualHost).forEach((name) => {
                    context.push(` ${name}="${value.content}"`);
                });
            }
            else {
                context.push(` ${prop.name}`);
            }
        }
        else {
            const { name } = prop;
            if (context.checkPropName &&
                !context.checkPropName(prop.name, prop, node)) {
                return;
            }
            if (name === 'on') {
                genOn(prop, node, context);
            }
            else {
                genDirectiveNode(prop, node, virtualHost, context);
            }
        }
    });
}
exports.genElementProps = genElementProps;
function genOn(prop, node, { push, event, isBuiltInComponent }) {
    if (!prop.arg) {
        return;
    }
    const arg = prop.arg.content;
    const exp = prop.exp;
    const modifiers = prop.modifiers;
    const name = (event?.format || uni_cli_shared_1.formatMiniProgramEvent)(arg, {
        isCatch: modifiers.includes('stop') || modifiers.includes('prevent'),
        isCapture: modifiers.includes('capture'),
        isComponent: (0, uni_cli_shared_1.isUserComponent)(node, { isBuiltInComponent }),
    });
    if (exp.isStatic) {
        push(` ${name}="${exp.content}"`);
    }
    else {
        push(` ${name}="{{${exp.content}}}"`);
    }
}
function genDirectiveNode(prop, node, virtualHost, context) {
    const { push, component } = context;
    if (prop.name === 'slot') {
        if (prop.arg) {
            const arg = prop.arg;
            if (arg.isStatic) {
                const slotName = (0, uni_shared_1.dynamicSlotName)(arg.content);
                // 非作用域默认插槽不生成 slot 属性
                if (slotName !== uni_shared_1.SLOT_DEFAULT_NAME) {
                    push(` slot="${slotName}"`);
                }
            }
            else {
                push(` slot="{{${arg.content}}}"`);
            }
        }
    }
    else if (prop.name === 'show') {
        let hiddenPropName = 'hidden';
        const value = `"{{!${prop.exp.content}}}"`;
        if ((0, uni_cli_shared_1.isUserComponent)(node, context)) {
            if (component && component.vShow) {
                hiddenPropName = component.vShow;
            }
            if (virtualHost) {
                // TODO use checkVirtualHostProps
                push(` ${uni_shared_1.VIRTUAL_HOST_HIDDEN}=${value}`);
            }
        }
        push(` ${hiddenPropName}=${value}`);
    }
    else if (prop.arg && prop.exp) {
        const arg = prop.arg.content;
        if (arg === utils_1.ATTR_ELEMENT_ID) {
            // 模板忽略生成 u-e，只需要 render 中生成
            return;
        }
        const exp = prop.exp.content;
        checkVirtualHostProps(arg, virtualHost).forEach((arg) => {
            // 组件作为根节点，virtualHostStyle="{{virtualHostStyle}}" 会产生警告 Setting data field "virtualHostStyle" to undefined is invalid.
            if (arg === uni_shared_1.VIRTUAL_HOST_STYLE && exp === uni_shared_1.VIRTUAL_HOST_STYLE) {
                push(` ${arg}="{{${exp} || ''}}"`);
            }
            else {
                push(` ${arg}="{{${exp}}}"`);
            }
        });
    }
    else {
        if (prop.name !== 'bind') {
            throw new Error(`unknown directive ` + JSON.stringify(prop));
        }
    }
}
